app/scripts/lib/app-start.js

```diff
@@ -304,66 +296,35 @@ define(function (require, exports, module) {
 
     initializeAuthenticationBroker () {
       if (! this._authenticationBroker) {
-        let assertionLibrary = this._assertionLibrary;
-        let iframeChannel = this._iframeChannel;
-        let metrics = this._metrics;
-        let oAuthClient = this._oAuthClient;
-        let relier = this._relier;
-        let session = Session;
-        let win = this._window;
-
-        if (this._isFxFirstrunV2()) {
-          this._authenticationBroker = new FxFirstrunV2AuthenticationBroker({
-            iframeChannel,
-            relier,
-            window: win
-          });
-        } else if (this._isFxFirstrunV1()) {
-          this._authenticationBroker = new FxFirstrunV1AuthenticationBroker({
-            iframeChannel,
-            relier,
-            window: win
-          });
-        } else if (this._isFxFennecV1()) {
-          this._authenticationBroker = new FxFennecV1AuthenticationBroker({
-            relier,
-            window: win
-          });
-        } else if (this._isFxDesktopV3()) {
-          this._authenticationBroker = new FxDesktopV3AuthenticationBroker({
-            relier,
-            window: win
-          });
-        } else if (this._isFxDesktopV2()) {
-          this._authenticationBroker = new FxDesktopV2AuthenticationBroker({
-            relier,
-            window: win
-          });
-        } else if (this._isFxDesktopV1()) {
-          this._authenticationBroker = new FxDesktopV1AuthenticationBroker({
-            relier,
-            window: win
-          });
-        } else if (this._isFxiOSV1()) {
-          this._authenticationBroker = new FxiOSV1AuthenticationBroker({
-            relier,
-            window: win
-          });
+        const args = {
+          assertionLibrary: this._assertionLibrary,
+          iframeChannel: this._iframeChannel,
+          metrics: this._metrics,
+          oAuthClient: this._oAuthClient,
+          relier: this._relier,
+          session: Session,
+          window: this._window
+        };
+
+        let context;
+        if (this._isFxDesktopV2()) {
+          if (this._isIframeContext()) {
+            context = Constants.FX_FIRSTRUN_V1_CONTEXT;
+          } else {
+            context = Constants.FX_DESKTOP_V2_CONTEXT;
+          }
         } else if (this._isOAuth()) {
-          this._authenticationBroker = new RedirectAuthenticationBroker({
-            assertionLibrary,
-            metrics,
-            oAuthClient,
-            relier,
-            session,
-            window: win
-          });
+          context = Constants.OAUTH_CONTEXT;
         } else {
-          this._authenticationBroker = new BaseAuthenticationBroker({
-            relier
-          });
+          context = this._getContext();
         }
 
+        const { Constructor, options } = authBrokers.get(context);
+        this._authenticationBroker = new Constructor(options.reduce((options, key) => {
```

Comment:
I think you can simplify lines 323-326 using [_.pick](http://underscorejs.org/#pick) with the [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator):

```js
this._authenticationBroker = new Constructor(_.pick(args, ...options));
```

Another option is to ignore options declaration and just pass all the args to all the constructors and let each module import the items it needs and ignore the rest. As it stands, imported options need to be declared twice to actually be imported, which is a potential foot gun when adding new options. If in the future we add a new `this._variableName = options.variableName` without adding a corresponding `variableName` in the options declaration, the variable won't be passed and won't be imported.

A third option is to go full declarative. You could declare the options block, then in the BaseAuthenticationBroker's constructor, import each of the items in `options`:

```js
constructor: function (options) {
  _.extend(this, options);
}
```

or for a bit of belt and suspenders to ensure no options are imported that aren't declared (assuming `options` is available on `this`):

```js
constructor: function (options) {
  _.extend(this, _.pick(options, ...this.options));
}
```

Doing the full declarative route would require some existing private variable names to lose their `_` prefix, so maybe a declarative way of declaring the "from=>to" would work:

```js
// declaring the options in the module
options: {
  assertionLibrary: '_assertionLibrary',
  oauthClient: `_oauthClient`
},
...
```

While you are doing such a major refactor, I'd go a bit further. First, the tiny back story - I screwed up when making the Authentication Brokers and didn't follow the Backbone Model convention for [constructor parameters](http://backbonejs.org/#Model-constructor) which is `new Model([attributes], [options])` where attributes are the first parameter, and options are the second. Instead, I only passed in `options` as `attributes` and as a result, all those options get imported into the model's attributes, which is Less Than Awesome. While digging in here, you could change all of the AuthBroker models to follow the Backbone convention.